File: projectforge-business/src/test/java/org/projectforge/fibu/RechnungDaoTest.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.fibu;

import static org.testng.AssertJUnit.assertEquals;
import static org.testng.AssertJUnit.fail;

import java.io.Serializable;
import java.math.BigDecimal;
import java.sql.Date;

import org.projectforge.business.fibu.RechnungDO;
import org.projectforge.business.fibu.RechnungDao;
import org.projectforge.business.fibu.RechnungFilter;
import org.projectforge.business.fibu.RechnungTyp;
import org.projectforge.business.fibu.RechnungsPositionDO;
import org.projectforge.framework.access.AccessException;
import org.projectforge.framework.i18n.UserException;
import org.projectforge.test.AbstractTestBase;
import org.springframework.beans.factory.annotation.Autowired;
import org.testng.annotations.Test;

public class RechnungDaoTest extends AbstractTestBase
{
  @Autowired
  private RechnungDao rechnungDao;

  private static int dbNumber = RechnungDao.START_NUMBER;

  @Test
  public void getNextNumber()
  {
    logon(TEST_FINANCE_USER);
    final RechnungDO rechnung1 = new RechnungDO();
    int number = rechnungDao.getNextNumber(rechnung1);
    rechnung1.setDatum(new Date(System.currentTimeMillis()));
    rechnung1.setFaelligkeit(new Date(System.currentTimeMillis()));
    rechnung1.setProjekt(initTestDB.addProjekt(null, 1, "foo"));
    try {
      rechnungDao.save(rechnung1);
      fail("Exception with wrong number should be thrown (no number given).");
    } catch (UserException ex) {
    }
    rechnung1.setNummer(number);
    rechnung1.addPosition(createPosition(1, "50.00", "0", "test"));
    Serializable id = rechnungDao.save(rechnung1);
    final RechnungDO rechnung1FromDb = rechnungDao.getById(id);
    assertEquals(dbNumber++, rechnung1FromDb.getNummer().intValue());

    final RechnungDO rechnung2 = new RechnungDO();
    rechnung2.setDatum(new Date(System.currentTimeMillis()));
    rechnung2.setNummer(number);
    try {
      rechnungDao.save(rechnung2);
      fail("Exception with wrong number should be thrown (does already exists).");
    } catch (UserException ex) {
    }
    number = rechnungDao.getNextNumber(rechnung2);
    rechnung2.setNummer(number + 1);
    rechnung2.setFaelligkeit(new Date(System.currentTimeMillis()));
    rechnung2.setProjekt(initTestDB.addProjekt(null, 1, "foo"));
    try {
      rechnungDao.save(rechnung2);
      fail("Exception with wrong number should be thrown (not continuously).");
    } catch (UserException ex) {
    }
    rechnung2.setNummer(number);
    rechnung2.addPosition(createPosition(1, "50.00", "0", "test"));
    id = rechnungDao.save(rechnung2);
    final RechnungDO rechnung2FromDb = rechnungDao.getById(id);
    assertEquals(dbNumber++, rechnung2FromDb.getNummer().intValue());

    final RechnungDO rechnung3 = new RechnungDO();
    rechnung3.setDatum(new Date(System.currentTimeMillis()));
    rechnung3.setTyp(RechnungTyp.GUTSCHRIFTSANZEIGE_DURCH_KUNDEN);
    rechnung3.addPosition(createPosition(1, "50.00", "0", "test"));
    rechnung3.setFaelligkeit(new Date(System.currentTimeMillis()));
    rechnung3.setProjekt(initTestDB.addProjekt(null, 1, "foo"));
    id = rechnungDao.save(rechnung3);
    final RechnungDO rechnung3FromDb = rechnungDao.getById(id);
    assertEquals(null, rechnung3FromDb.getNummer());
    dbNumber++; // Needed for getNextNumber test;
  }

  @Test
  public void checkAccess()
  {
    logon(TEST_FINANCE_USER);
    RechnungDO rechnung = new RechnungDO();
    int number = rechnungDao.getNextNumber(rechnung);
    rechnung.setDatum(new Date(System.currentTimeMillis()));
    rechnung.setFaelligkeit(new Date(System.currentTimeMillis()));
    rechnung.setProjekt(initTestDB.addProjekt(null, 1, "foo"));
    rechnung.setNummer(number);

    rechnung.addPosition(createPosition(2, "100.50", "0.19", "test"));
    rechnung.addPosition(createPosition(1, "50.00", "0", "test"));
    assertEquals("289.19", String.valueOf(rechnung.getGrossSum().setScale(2)));
    Serializable id = rechnungDao.save(rechnung);
    dbNumber++;
    rechnung = rechnungDao.getById(id);

    logon(TEST_CONTROLLING_USER);
    rechnungDao.getById(id);
    checkNoWriteAccess(id, rechnung, "Controlling");

    logon(TEST_USER);
    checkNoAccess(id, rechnung, "Other");

    logon(TEST_PROJECT_MANAGER_USER);
    checkNoAccess(id, rechnung, "Project manager");

    logon(TEST_ADMIN_USER);
    checkNoAccess(id, rechnung, "Admin ");
  }

  private void checkNoAccess(Serializable id, RechnungDO rechnung, String who)
  {
    try {
      RechnungFilter filter = new RechnungFilter();
      rechnungDao.getList(filter);
      fail("AccessException expected: " + who + " users should not have select list access to invoices.");
    } catch (AccessException ex) {
      // OK
    }
    try {
      rechnungDao.getById(id);
      fail("AccessException expected: " + who + " users should not have select access to invoices.");
    } catch (AccessException ex) {
      // OK
    }
    checkNoHistoryAccess(id, rechnung, who);
    checkNoWriteAccess(id, rechnung, who);
  }

  private void checkNoHistoryAccess(Serializable id, RechnungDO rechnung, String who)
  {
    assertEquals(who + " users should not have select access to history of invoices.",
        rechnungDao.hasLoggedInUserHistoryAccess(false), false);
    try {
      rechnungDao.hasLoggedInUserHistoryAccess(true);
      fail("AccessException expected: " + who + " users should not have select access to history of invoices.");
    } catch (AccessException ex) {
      // OK
    }
    assertEquals(who + " users should not have select access to history of invoices.",
        rechnungDao.hasLoggedInUserHistoryAccess(rechnung, false), false);
    try {
      rechnungDao.hasLoggedInUserHistoryAccess(rechnung, true);
      fail("AccessException expected: " + who + " users should not have select access to history of invoices.");
    } catch (AccessException ex) {
      // OK
    }
  }

  private void checkNoWriteAccess(Serializable id, RechnungDO rechnung, String who)
  {
    try {
      RechnungDO re = new RechnungDO();
      int number = rechnungDao.getNextNumber(re);
      re.setDatum(new Date(System.currentTimeMillis()));
      re.setNummer(number);
      rechnungDao.save(re);
      fail("AccessException expected: " + who + " users should not have save access to invoices.");
    } catch (AccessException ex) {
      // OK
    }
    try {
      rechnung.setBemerkung(who);
      rechnungDao.update(rechnung);
      fail("AccessException expected: " + who + " users should not have update access to invoices.");
    } catch (AccessException ex) {
      // OK
    }
  }

  private RechnungsPositionDO createPosition(final int menge, final String einzelNetto, final String vat,
      final String text)
  {
    final RechnungsPositionDO pos = new RechnungsPositionDO();
    pos.setMenge(new BigDecimal(menge));
    pos.setEinzelNetto(new BigDecimal(einzelNetto));
    pos.setVat(new BigDecimal(vat));
    pos.setText(text);
    return pos;
  }

}

File: projectforge-business/src/test/java/org/projectforge/business/fibu/RechnungDaoTest.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2024 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.fibu

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.projectforge.common.i18n.UserException
import org.projectforge.framework.access.AccessException
import org.projectforge.test.AbstractTestBase
import org.springframework.beans.factory.annotation.Autowired
import java.io.Serializable
import java.math.BigDecimal
import java.time.LocalDate

class RechnungDaoTest : AbstractTestBase() {
    @Autowired
    private lateinit var rechnungDao: RechnungDao

    @Test
    fun testNextNumber() {
        persistenceService.runInTransaction { _ ->
            var dbNumber = rechnungDao.nextNumber
            logon(TEST_FINANCE_USER)
            val rechnung1 = RechnungDO()
            var number = rechnungDao.getNextNumber(rechnung1)
            rechnung1.datum = LocalDate.now()
            rechnung1.faelligkeit = LocalDate.now()
            rechnung1.projekt = initTestDB.addProjekt(null, 1, "foo")
            try {
                rechnungDao.insert(rechnung1)
                Assertions.fail("Exception with wrong number should be thrown (no number given).")
            } catch (ex: UserException) {
                // Expected
            }
            rechnung1.nummer = number
            rechnung1.addPosition(createPosition(1, "50.00", "0", "test"))
            var id: Serializable? = rechnungDao.insert(rechnung1)
            val rechnung1FromDb = rechnungDao.find(id, attached = true) // Attached is important, otherwise deadlock.
            Assertions.assertEquals(dbNumber++, rechnung1FromDb!!.nummer)

            val rechnung2 = RechnungDO()
            rechnung2.datum = LocalDate.now()
            rechnung2.nummer = number
            try {
                rechnungDao.insert(rechnung2)
                Assertions.fail<Any>("Exception with wrong number should be thrown (does already exists).")
            } catch (ex: UserException) {
                // Expected.
            }
            number = rechnungDao.getNextNumber(rechnung2)
            rechnung2.nummer = number + 1
            rechnung2.faelligkeit = LocalDate.now()
            rechnung2.projekt = initTestDB.addProjekt(null, 1, "foo")
            try {
                rechnungDao.insert(rechnung2)
                Assertions.fail<Any>("Exception with wrong number should be thrown (not continuously).")
            } catch (ex: UserException) {
                // OK
            }
            rechnung2.nummer = number
            rechnung2.addPosition(createPosition(1, "50.00", "0", "test"))
            id = rechnungDao.insert(rechnung2)
            val rechnung2FromDb = rechnungDao.find(id, attached = true) // Attached is important, otherwise deadlock.
            Assertions.assertEquals(dbNumber, rechnung2FromDb!!.nummer)

            val rechnung3 = RechnungDO()
            rechnung3.datum = LocalDate.now()
            rechnung3.typ = RechnungTyp.GUTSCHRIFTSANZEIGE_DURCH_KUNDEN
            rechnung3.addPosition(createPosition(1, "50.00", "0", "test"))
            rechnung3.faelligkeit = LocalDate.now()
            rechnung3.projekt = initTestDB.addProjekt(null, 1, "foo")
            id = rechnungDao.insert(rechnung3)
            val rechnung3FromDb = rechnungDao.find(id, attached = true) // Attached is important, otherwise deadlock.
            Assertions.assertNull(rechnung3FromDb!!.nummer)
        }
    }

    @Test
    fun checkAccess() {
        lateinit var rechnung: RechnungDO
        lateinit var id: Serializable
        persistenceService.runInTransaction<Any?> { context ->
            logon(TEST_FINANCE_USER)
            rechnung = RechnungDO()
            val number = rechnungDao.getNextNumber(rechnung)
            rechnung.datum = LocalDate.now()
            rechnung.faelligkeit = LocalDate.now()
            rechnung.projekt = initTestDB.addProjekt(null, 1, "foo")
            rechnung.nummer = number

            rechnung.addPosition(createPosition(2, "100.50", "0.19", "test"))
            rechnung.addPosition(createPosition(1, "50.00", "0", "test"))
            Assertions.assertEquals("289.19", rechnung.grossSum.setScale(2).toString())
            id = rechnungDao.insert(rechnung)
            rechnung = rechnungDao.find(id, attached = true)!! // Attached is important, otherwise deadlock.
        }
        persistenceService.runInTransaction { context ->
            logon(TEST_CONTROLLING_USER)
            rechnungDao.find(id, attached = true) // Attached is important, otherwise deadlock.
            checkNoWriteAccess(rechnung, "Controlling")

            logon(TEST_USER)
            checkNoAccess(id, rechnung, "Other")

            logon(TEST_PROJECT_MANAGER_USER)
            checkNoAccess(id, rechnung, "Project manager")

            logon(TEST_ADMIN_USER)
            checkNoAccess(id, rechnung, "Admin ")
            null
        }
    }

    private fun checkNoAccess(id: Serializable, rechnung: RechnungDO, who: String) {
        try {
            val filter = RechnungFilter()
            rechnungDao.select(filter)
            Assertions.fail<Any>("AccessException expected: $who users should not have select list access to invoices.")
        } catch (ex: AccessException) {
            // OK
        }
        try {
            rechnungDao.find(id, attached = true) // Attached is important, otherwise deadlock.
            Assertions.fail<Any>("AccessException expected: $who users should not have select access to invoices.")
        } catch (ex: AccessException) {
            // OK
        }
        checkNoHistoryAccess(rechnung, who)
        checkNoWriteAccess(rechnung, who)
    }

    private fun checkNoHistoryAccess(rechnung: RechnungDO, who: String) {
        Assertions.assertFalse(
            rechnungDao.hasLoggedInUserHistoryAccess(false),
            "$who users should not have select access to history of invoices."
        )
        try {
            rechnungDao.hasLoggedInUserHistoryAccess(true)
            Assertions.fail<Any>("AccessException expected: $who users should not have select access to history of invoices.")
        } catch (ex: AccessException) {
            // OK
        }
        Assertions.assertFalse(
            rechnungDao.hasLoggedInUserHistoryAccess(rechnung, false),
            "$who users should not have select access to history of invoices."
        )
        try {
            rechnungDao.hasLoggedInUserHistoryAccess(rechnung, true)
            Assertions.fail<Any>("AccessException expected: $who users should not have select access to history of invoices.")
        } catch (ex: AccessException) {
            // OK
        }
    }

    private fun checkNoWriteAccess(rechnung: RechnungDO, who: String) {
        try {
            val re = RechnungDO()
            val number = rechnungDao.getNextNumber(re)
            re.datum = LocalDate.now()
            re.nummer = number
            rechnungDao.insert(re)
            Assertions.fail<Any>("AccessException expected: $who users should not have save access to invoices.")
        } catch (ex: AccessException) {
            // OK
        }
        try {
            rechnung.bemerkung = who
            rechnungDao.update(rechnung)
            Assertions.fail<Any>("AccessException expected: $who users should not have update access to invoices.")
        } catch (ex: AccessException) {
            // OK
        }
    }

    private fun createPosition(
        menge: Int, einzelNetto: String, vat: String,
        text: String
    ): RechnungsPositionDO {
        val pos = RechnungsPositionDO()
        pos.menge = BigDecimal(menge)
        pos.einzelNetto = BigDecimal(einzelNetto)
        pos.vat = BigDecimal(vat)
        pos.text = text
        return pos
    }
}
